### PR TITLE
Rewrite feature registry with new Pythonic APIs and add tagging support

### PR DESCRIPTION
**Pull request by xiaoyongzhu**
This PR focus on the following goals with minimal end user facing impact:
1. shift the implementation from HOCON based parsing to python API parsing, which is more stable and more scalable.
2. Fix the hierarchy issue. Previously source is on project level. with this fix, source should be on anchor level.
3. Derived feature support. Now the input features for derived features can be parsed.
4. Supporting adding tags for all the Feathr entities, including project, anchor, anchor feature, derived feature, source, to allow customization.
5. helper functions to facilitate testing & migration
6. feathr entity search support ,which facilitate future development work
7. Remove undocumented functionalities that needs further discussion (client.get_features)

image

image

### CODE REVIEW

File: feathr_project/setup.py [hunk 1]
```diff
@@ -40,6 +40,7 @@
         "python-snappy",
         "deltalake",
         "google>=3.0.0",
+        "graphlib_backport",
```
Comment (jaymo001) [inline feathr_project/setup.py:43]: The license seems not compatible with Apache 2?

File: feathr_project/feathr/_feature_registry.py [hunk 2]
```diff
@@ -646,9 +554,12 @@ def save_to_feature_config_from_context(self, anchor_list, derived_feature_list,
         """Save feature definition within the workspace into HOCON feature config files from current context, rather than reading from python files"""
         repo_definitions = self._extract_features_from_context(
             anchor_list, derived_feature_list, local_workspace_dir)
-        self._save_request_feature_config(repo_definitions, local_workspace_dir)
-        self._save_anchored_feature_config(repo_definitions, local_workspace_dir)
-        self._save_derived_feature_config(repo_definitions, local_workspace_dir)
+        self._save_request_feature_config(
+            repo_definitions, local_workspace_dir)
```
Comment (jaymo001) [inline feathr_project/feathr/_feature_registry.py:558]: nit: format is inconsistent with rest of the code

File: feathr_project/feathr/_feature_registry.py [hunk 3]
```diff
@@ -725,7 +640,8 @@ def register_features(self, workspace_path: Optional[Path] = None):
 
         # register feature types each time when we register features.
         self._register_feathr_feature_types()
-        self._read_config_from_workspace(workspace_path)
+        self._parse_features_from_context(
```
Comment (jaymo001) [inline feathr_project/feathr/_feature_registry.py:643]: The input 'from_context' in unused

File: feathr_project/feathr/_feature_registry.py [hunk 4]
```diff
@@ -734,76 +650,210 @@ def register_features(self, workspace_path: Optional[Path] = None):
...
+            logger.info("{} feathr entities deleted", batch_delte_size)
+
+    def _get_registry_client(self):
         """
         Return a client object and users can operate more on it (like doing search)
         """
         return self.purview_client
 
-    def list_registered_features(self, project_name: str = None) -> List[str]:
+    def list_registered_features(self, project_name: str = None, limit=50, starting_offset=0,) -> List[str]:
```
Comment (jaymo001) [inline feathr_project/feathr/_feature_registry.py:712]: extra trailing ','

File: feathr_project/test/test_feature_registry.py [hunk 5]
```diff
@@ -3,18 +3,22 @@
...
+import time
 import pytest
 
 def test_feathr_register_features_e2e():
     runner = CliRunner()
     with runner.isolated_filesystem():
         runner.invoke(init, [])
-        client = basic_test_setup(
+        client = registry_test_setup(
             "./feathr_user_workspace/feathr_config.yaml")
```
Comment (hangfei) [inline feathr_project/test/test_feature_registry.py:15]: use test_user_workspace

File: feathr_project/setup.py (hunk 1 above)
Comment (hangfei) [inline feathr_project/setup.py:43]: why is this backport?

File: feathr_project/feathr/anchor.py [hunk 6]
```diff
@@ -15,15 +15,19 @@ class FeatureAnchor:
     Attributes:
         name: Unique name of the anchor.
         source: data source that the features are anchored to. Should be either of `INPUT_CONTEXT` or `feathr.source.Source`
-        features: list of features within this anchor.
+        features: list of features within this anchor. 
+        registry_tags: A dict of (str, str) that you can pass to feature registry for customization. For example, you can use `registry_tags` to indicate whether this anchor is deprecated or not, etc.
```
Comment (hangfei) [inline feathr_project/feathr/anchor.py:19]: why is it A dict of (str, str)? Provide example

File: feathr_project/feathr/feature_derivations.py [hunk 7]
```diff
@@ -16,15 +16,18 @@ class DerivedFeature(FeatureBase):
         key: All features with corresponding keys that this derived feature depends on
         input_features: features that this derived features depends on
         transform: transformation that produces the derived feature value, based on the input_features
+        registry_tags: A dict of (str, str) that you can pass to feature registry for customization. For example, you can use `registry_tags` to indicate feature description, whether this feature is deprecated or not, last refreshed time, etc.
```
Comment (hangfei) [inline feathr_project/feathr/feature_derivations.py:19]: it's better to have a class for registry_tags so you don't have to keep repeating the documenation. and also can provide some methods for registry tags.

File: feathr_project/feathr/constants.py [hunk 8]
```diff
@@ -6,3 +6,21 @@
...
+
+TYPEDEF_SOURCE=f'feathr_source_{REGISTRY_TYPEDEF_VERSION}'
+TYPEDEF_FEATHR_PROJECT=f'feathr_workspace_{REGISTRY_TYPEDEF_VERSION}'
+TYPEDEF_DERIVED_FEATURE=f'feathr_derived_feature_{REGISTRY_TYPEDEF_VERSION}'
+TYPEDEF_ANCHOR=f'feathr_anchor_{REGISTRY_TYPEDEF_VERSION}'
+TYPEDEF_ANCHOR_FEATURE=f'feathr_anchor_feature_{REGISTRY_TYPEDEF_VERSION}'
+
+TYPEDEF_ARRAY_ANCHOR=f"array<feathr_anchor_{REGISTRY_TYPEDEF_VERSION}>"
+TYPEDEF_ARRAY_DERIVED_FEATURE=f"array<feathr_derived_feature_{REGISTRY_TYPEDEF_VERSION}>"
+TYPEDEF_ARRAY_ANCHOR_FEATURE=f"array<feathr_anchor_feature_{REGISTRY_TYPEDEF_VERSION}>"
```
Comment (hangfei) [inline feathr_project/feathr/constants.py:25]: i don't actually this used anywhere. so i am confused why we need this and how it's used.

**Review by hangfei — APPROVED** [review]

**Review by jaymo001 — APPROVED** [review]

**Review by hangfei — APPROVED** [review]

**Pull request merged**